### PR TITLE
Move vault to pre-deploy and split managedIdentity 

### PR DESCRIPTION
**Pull request by mjudeikis**
1. Moves managedIdentity to separate deployment
2. Moves vault to preDeploy
3. Everything else

At the same time, splits generators even more to contain all configurable in the `templates` files and all abstract logic in the `generators`. Simplifies the generation model.

### CODE REVIEW

**Comment by julienstroheker** [comment]
Docs need to be updated

File: hack/gendeploy/gendeploy.go [hunk 1]
```diff
@@ -8,22 +8,16 @@ import (
 )
 
 func run() error {
-	err := generator.GenerateRPTemplates()
+	// dev artifacts
+	gen := generator.New(false)
+	err := gen.Artifacts()
```
[the remark this answers is not in the document]
Comment (mjudeikis) [reply hack/gendeploy/gendeploy.go]: Could. But is all about readability, not shorted script.

File: pkg/deploy/deploy.go [hunk 2]
```diff
@@ -87,11 +87,57 @@ func (d *deployer) PreDeploy(ctx context.Context) (rpServicePrincipalID string,
...
+		return "", err
+	}
+
+	return rpServicePrincipalID, d.deployPreDeploy(ctx, rpServicePrincipalID)
+}
+
+func (d *deployer) deployPreDeploy(ctx context.Context, rpServicePrincipalID string) (err error) {
+	deploymentName := "rp-production-predeploy"
+	_, err = d.deployments.Get(ctx, d.resourceGroup, deploymentName)
+	if isDeploymentNotFoundError(err) {
```
Comment (jim-minter) [inline pkg/deploy/deploy.go]: `if !isDeploymentNotFoundError(err) { return err }` saves you an indent --> better; you can then also remove the `var b []byte` below and do `b, err := Asset(...)`

File: pkg/deploy/deploy.go [hunk 3]
```diff
@@ -113,7 +159,11 @@ func (d *deployer) PreDeploy(ctx context.Context) (rpServicePrincipalID string,
 			return "", err
 		}
 
-		deployment, err = d.deployments.Get(ctx, d.resourceGroup, deploymentName)
+		deployment, err := d.deployments.Get(ctx, d.resourceGroup, deploymentName)
+		if err != nil {
+			return "", err
+		}
+		return deployment.Properties.Outputs.(map[string]interface{})["rpServicePrincipalId"].(map[string]interface{})["value"].(string), nil
```
Comment (jim-minter) [inline pkg/deploy/deploy.go]: this is not an improvement.  Perhaps putting all of this scope into a separate function, say "_deployManageIdentity" will help.

File: pkg/deploy/generator/util.go [hunk 4]
```diff
@@ -3,9 +3,48 @@ package generator
 // Copyright (c) Microsoft Corporation.
 // Licensed under the Apache License 2.0.
 
+import (
```
Comment (jim-minter) [inline pkg/deploy/generator/util.go]: I am -1000 on util.go files.  Each of the functions below is called exclusively from one .go file.  Move each function in question into that .go file please.  Including max.

File: pkg/deploy/generator/templates.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	mgmtkeyvault "github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2016-10-01/keyvault"
+
+	"github.com/Azure/ARO-RP/pkg/util/arm"
+)
+
+func (g *generator) _managedIdentityTemplate() *arm.Template {
```
Comment (jim-minter) [inline pkg/deploy/generator/templates.go]: please remove the leading `_` from all of these functions.

File: pkg/deploy/generator/templates.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+		},
+		"databaseName": {
+			Type: "string",
+		},
+	}
+
+	return t
+}
+
+func (g *generator) _rpParametersTemplate() *arm.Parameters {
```
Comment (jim-minter) [inline pkg/deploy/generator/templates.go]: this isn't returning a template - call it something like `rpParameters` ?  (please do the same for the others)

File: pkg/deploy/generator/templates.go [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+			p.MaxLength = 24 - max(len(kvClusterSuffix), len(kvServiceSuffix))
+		}
+		t.Parameters[param] = p
+	}
+
+	t.Resources = append(t.Resources,
+		g.securityGroupRP(),
+		g.securityGroupPE(),
+		// clustersKeyvault must preceed serviceKeyvault due to terrible
+		// bytes.Replace below
```
Comment (jim-minter) [inline pkg/deploy/generator/templates.go]: update the comment please

File: pkg/deploy/generator/templates.go [hunk 8]
```diff
@@ -0,0 +1,221 @@
+package generator
```
Comment (jim-minter) [inline pkg/deploy/generator/templates.go]: this file came out well, didn't it!

File: pkg/deploy/generator/generators.go [hunk 9]
```diff
@@ -32,290 +29,163 @@ var (
...
+	// these are not exposed, but this list acts more as a documentation
+	managedIdentityTemplate(outputFile string) error
+	rpTemplate(outputFile string) error
+	databaseTemplate(outputFile string) error
+	preDeployTemplates(outputFile string) error
+
+	rpParametersTemplate(outputFile string) error
+	rpPreDeployParametersTemplate(outputFile string) error
+
+	sharedDevelopmentEnvTemplate(outputFile string) error
```
Comment (jim-minter) [inline pkg/deploy/generator/generators.go]: ick.  I think that function is being performed pretty clearly by Artifacts() - please remove this

File: pkg/deploy/generator/generators.go [hunk 10]
```diff
@@ -32,290 +29,163 @@ var (
...
-			},
-			"databaseName": {
-				Type: "string",
-			},
-		},
-	}
+}
 
-	g := newGenerator(false)
+func (g *generator) managedIdentityTemplate(output string) error {
```
Comment (jim-minter) [inline pkg/deploy/generator/generators.go]: from here down is very repetitive, and I don't think the `if g.production`s add value - it's all encoded directly above.  How about two functions, one for parameters and one for templates.  Either their signature is like `writeTemplate(f func() *arm.Template, filename string) error` or it's like `writeTemplate(t *arm.Template, filename string) error`.

File: pkg/deploy/deploy.go (hunk 3 above)
Comment (mjudeikis) [reply pkg/deploy/deploy.go]: didnt understood. It's already in similar function...

File: pkg/deploy/generator/templates.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/deploy/generator/templates.go]: If I do this I have to rename each function in `generators.go` with different name:
```
sharedDevelopmentEnvTemplate(file){
   _sharedDevelopmentEnvTemplate()
   write file
```
If I rename those with `generateSharedDevelopmentEnvTemplate(file)` I get `generate.generateSharedDevelopmentEnvTemplate(file)` which is repetetive. I can't come up with new name for these which would not be repetetetive and nice, simple.

File: pkg/deploy/generator/generators.go (hunk 9 above)
Comment (mjudeikis) [reply pkg/deploy/generator/generators.go]: I already got feedback that things like this help people to understand code easier and faster. I can remove it, but I think we basically making code harder to understand and raising entry bar :/

File: pkg/deploy/generator/generators.go (hunk 10 above)
Comment (mjudeikis) [reply pkg/deploy/generator/generators.go]: `if.production` prevents from fool prof errors. 
Functions - could be done, but first `_function` naming need to be decided, as as stated above - I need a suggestion if you don't like current naming :/

**Pull request merged**